### PR TITLE
[BEAM-2020] Add additional CloudObject Translators

### PR DESCRIPTION
**Pull request by tgroh**
Be sure to do all of the following to help us incorporate your contribution
quickly and easily:

 - [x] Make sure the PR title is formatted like:
   `[BEAM-<Jira issue #>] Description of pull request`
 - [x] Make sure tests pass via `mvn clean verify`. (Even better, enable
       Travis-CI on your fork and ensure the whole test matrix passes).
 - [x] Replace `<Jira issue #>` in the title with the actual Jira issue
       number, if there is one.
 - [x] If this contribution is large, please file an Apache
       [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

---

### CODE REVIEW

**Comment by tgroh** [comment]
R: @lukecwik

File: runners/google-cloud-dataflow-java/src/main/java/org/apache/beam/runners/dataflow/util/DefaultCoderCloudObjectTranslatorRegistrar.java [hunk 1]
```diff
@@ -65,7 +67,15 @@
           CloudObjectTranslators.stream(),
           CloudObjectTranslators.pair(),
           CloudObjectTranslators.windowedValue(),
-          CloudObjectTranslators.custom());
+          CloudObjectTranslators.custom(),
+          CloudObjectTranslators.iterableLike(ListCoder.class),
+          CloudObjectTranslators.iterableLike(SetCoder.class),
```
Comment (lukecwik) [inline runners/google-cloud-dataflow-java/src/main/java/org/apache/beam/runners/dataflow/util/DefaultCoderCloudObjectTranslatorRegistrar.java]: Missing CollectionCoder.class

File: runners/google-cloud-dataflow-java/src/main/java/org/apache/beam/runners/dataflow/util/CloudObjectTranslators.java [hunk 2]
```diff
@@ -372,4 +381,169 @@ public String cloudObjectClassName() {
...
+            base, PropertyNames.CO_GBK_RESULT_SCHEMA, toCloudObject(target.getSchema()));
+        return addComponents(base, Collections.singletonList(target.getUnionCoder()));
+      }
+
+      private CloudObject toCloudObject(CoGbkResultSchema schema) {
+        CloudObject result = CloudObject.forClass(CoGbkResultSchema.class);
+        List<CloudObject> tags = new ArrayList<>(schema.getTupleTagList().size());
+        for (TupleTag<?> tag : schema.getTupleTagList().getAll()) {
+          CloudObject tagCloudObject = CloudObject.forClass(TupleTag.class);
+          Structs.addString(tagCloudObject, PropertyNames.VALUE, tag.getId());
```
Comment (lukecwik) [inline runners/google-cloud-dataflow-java/src/main/java/org/apache/beam/runners/dataflow/util/CloudObjectTranslators.java]: We lost the IS_GENERATED but its an unused property within Dataflow.

File: runners/google-cloud-dataflow-java/src/main/java/org/apache/beam/runners/dataflow/util/DefaultCoderCloudObjectTranslatorRegistrar.java [hunk 3]
```diff
@@ -65,7 +67,15 @@
...
+          CloudObjectTranslators.map(),
+          CloudObjectTranslators.nullable(),
```
Comment (lukecwik) [inline runners/google-cloud-dataflow-java/src/main/java/org/apache/beam/runners/dataflow/util/DefaultCoderCloudObjectTranslatorRegistrar.java]: Add tests for the new cases

File: runners/google-cloud-dataflow-java/src/main/java/org/apache/beam/runners/dataflow/util/DefaultCoderCloudObjectTranslatorRegistrar.java (hunk 1 above)
Comment (tgroh) [inline runners/google-cloud-dataflow-java/src/main/java/org/apache/beam/runners/dataflow/util/DefaultCoderCloudObjectTranslatorRegistrar.java]: Added

File: runners/google-cloud-dataflow-java/src/main/java/org/apache/beam/runners/dataflow/util/DefaultCoderCloudObjectTranslatorRegistrar.java (hunk 3 above)
Comment (tgroh) [inline runners/google-cloud-dataflow-java/src/main/java/org/apache/beam/runners/dataflow/util/DefaultCoderCloudObjectTranslatorRegistrar.java]: Added

File: runners/google-cloud-dataflow-java/src/main/java/org/apache/beam/runners/dataflow/util/CloudObjectTranslators.java (hunk 2 above)
Comment (tgroh) [inline runners/google-cloud-dataflow-java/src/main/java/org/apache/beam/runners/dataflow/util/CloudObjectTranslators.java]: ACK.

File: runners/google-cloud-dataflow-java/src/test/java/org/apache/beam/runners/dataflow/util/CloudObjectsTest.java [hunk 4]
```diff
@@ -112,7 +121,25 @@ public void defaultCodersIncludesCustomCoder() {
                       KvCoder.of(VarLongCoder.of(), ByteArrayCoder.of()),
                       IntervalWindow.getCoder()))
               .add(VarLongCoder.of())
-              .add(ByteArrayCoder.of());
+              .add(ByteArrayCoder.of())
```
Comment (lukecwik) [inline runners/google-cloud-dataflow-java/src/test/java/org/apache/beam/runners/dataflow/util/CloudObjectsTest.java]: Can you move ByteArrayCoder.of up so this list and the list in DefaultCoderRegistrars is in the same order, makes it easier to review.

Comment (tgroh) [inline runners/google-cloud-dataflow-java/src/test/java/org/apache/beam/runners/dataflow/util/CloudObjectsTest.java]: Done.

The `DefaultsPresentTest` should fail if there's any coder with a custom serializer/deserializer that isn't exercised by the `DefaultCoders` test.

**Comment by lukecwik** [comment]
retest this please

**Comment by lukecwik** [comment]
retest this please

File: runners/google-cloud-dataflow-java/src/test/java/org/apache/beam/runners/dataflow/util/CloudObjectsTest.java [hunk 5]
```diff
@@ -113,10 +122,28 @@ public void defaultCodersIncludesCustomCoder() {
                   WindowedValue.getFullCoder(
                       KvCoder.of(VarLongCoder.of(), ByteArrayCoder.of()),
                       IntervalWindow.getCoder()))
-              .add(VarLongCoder.of())
               .add(ByteArrayCoder.of())
+              .add(VarLongCoder.of())
               .add(SerializableCoder.of(Record.class))
               .add(AvroCoder.of(Record.class));
```
Comment (lukecwik) [inline runners/google-cloud-dataflow-java/src/test/java/org/apache/beam/runners/dataflow/util/CloudObjectsTest.java]: Remove extraneous ;

**Pull request merged**